File: harmony_app/entry/src/main/ets/utils/EncodingUtil.ts
import { fileIo } from '@kit.CoreFileKit';
import { util } from '@kit.ArkTS';

/**
 * 文件编码探测工具，对齐安卓端 EncodingUtil + iOS 端 EncodingUtil.decodeStrict。
 *
 * 探测规则：
 * 1. 先检查 BOM（UTF-8、UTF-16LE、UTF-16BE）；
 * 2. 无 BOM 时按字节校验是否合法 UTF-8，且对采样尾部被截断的不完整多字节序列
 *    视作合法前缀（对齐 iOS/安卓，避免采样边界误判）；
 * 3. 关键修复：引入 decodeStrict（移植自 iOS）做 CJK 占比评分 + 异常字符惩罚，
 *    纠正「GBK 文件被误判为合法 UTF-8 → 用 UTF-8 解码成乱码（损坏文件）」的问题；
 *    并以 GB18030 兜底（中文 txt 常见编码，GBK 超集），失败再回退 GBK。
 *
 * 仅在“深度扫描”模式下调用 detectEncodingName，对齐安卓端 deep 模式才检测编码的语义。
 */
export class EncodingUtil {
  private static readonly SAMPLE_BYTES: number = 16 * 1024;

  /** 探测文件编码显示名：UTF-8 / UTF-16LE / UTF-16BE / GB18030 / GBK。 */
  public static detectEncodingName(uri: string): string {
    let fd: number = 0;
    try {
      const file = fileIo.openSync(uri, fileIo.OpenMode.READ_ONLY);
      fd = file.fd;
      const buf: ArrayBuffer = new ArrayBuffer(EncodingUtil.SAMPLE_BYTES);
      const len: number = fileIo.readSync(fd, buf, { offset: 0, length: EncodingUtil.SAMPLE_BYTES });
      const bytes: Uint8Array = new Uint8Array(buf.slice(0, len));
      return EncodingUtil.detect(bytes);
    } catch (e) {
      return 'UTF-8';
    } finally {
      if (fd !== 0) {
        try {
          fileIo.closeSync(fd);
        } catch (e) {
          // 忽略关闭异常
        }
      }
    }
  }

  private static detect(bytes: Uint8Array): string {
    const len: number = bytes.length;
    if (len >= 3 && bytes[0] === 0xEF && bytes[1] === 0xBB && bytes[2] === 0xBF) {
      return 'UTF-8';
    }
    if (len >= 2 && bytes[0] === 0xFF && bytes[1] === 0xFE) {
      return 'UTF-16LE';
    }
    if (len >= 2 && bytes[0] === 0xFE && bytes[1] === 0xFF) {
      return 'UTF-16BE';
    }
    // 关键修复：用 decodeStrict 严格解码并评分，对齐 iOS：
    // 若 UTF-8 严格解码后 CJK 占比达标且异常字符少，则判 UTF-8；
    // 否则用 GB18030 严格解码评分，谁更"像中文文本"用谁，最后兜底 GBK。
    return EncodingUtil.decodeStrict(bytes);
  }

  /**
   * 严格解码并自动选编码（移植自 iOS decodeStrict，对齐安卓 Charset 严格解码语义）。
   *
   * 修复点：原生 util.TextDecoder 解码非法 UTF-8 时会静默丢弃/替换字节，
   * 造成 GBK 文件被当成 UTF-8 解码后满满乱码（用户感知为"损坏的文件"）。
   * 这里通过 CJK 占比评分 + 异常字符惩罚，在 UTF-8 与 GB18030 之间择优：
   * 数字符层面占比，而非简单"能否解码"。
   *
   * @param bytes 采样字节（建议 ≥ 1KB，越长判定越准）
   * @param prefer 可选，外部已确定主候选（如已知含 BOM 的 UTF-8）；为空则自动评分
   * @returns 显示名：'UTF-8' / 'GB18030' / 'GBK'
   */
  public static decodeStrict(bytes: Uint8Array, prefer?: string): string {
    if (!bytes || bytes.length === 0) {
      return prefer && prefer.length > 0 ? prefer : 'UTF-8';
    }

    // 候选口径（对齐安卓 GB18030 优先）：Utf8 → Gb18030 → Gbk
    const utfText: string = EncodingUtil.decodeBytes(bytes, 'utf-8');
    const gbText: string = EncodingUtil.decodeBytes(bytes, 'gb18030');
    const gbkText: string = EncodingUtil.decodeBytes(bytes, 'gbk');

    if (!utfText) {
      // UTF-8 严格解码完全失败（含非法序列），直接用 GB18030（含 GBK 字符集）
      return gbkText && gbkText.length > 0 ? 'GB18030' : 'GBK';
    }

    const utfScore: number = EncodingUtil.scoreText(utfText);

    // 已指定主候选：该候选解码成功且评分达标即采用；否则落回自动评分
    if (prefer && prefer.length > 0) {
      if (prefer.toUpperCase() === 'UTF-8') {
        if (utfScore >= 0.4) return 'UTF-8';
      } else {
        const pText: string = EncodingUtil.decodeBytes(bytes, prefer.toLowerCase());
        if (pText && EncodingUtil.scoreText(pText) > utfScore) {
          return prefer.toUpperCase();
        }
      }
    }

    // 自动评分：GB18030 与 UTF-8 二选一
    const gbScore: number = EncodingUtil.scoreText(gbText);
    if (gbScore > utfScore && gbScore >= 0.4) {
      return 'GB18030';
    }
    return 'UTF-8';
  }

  /**
   * 文本"像中文文本"的评分（移植自 iOS cjkScore）：
   * - CJK 统一表意文字（含扩展区）占比越高越好；
   * - 异常字符（U+FFFD 替换符、孤立控制字符、孤立代理项）越多越像损坏乱码，扣分。
   */
  private static scoreText(text: string): number {
    if (!text || text.length === 0) return 0;
    let cjk: number = 0;
    let abnormal: number = 0;
    const n: number = text.length;
    for (let i = 0; i < n; i++) {
      const cp: number = text.charCodeAt(i);
      // 先归一化到码点（处理 UTF-16 代理对）
      let code: number = cp;
      if (cp >= 0xD800 && cp <= 0xDBFF && i + 1 < n) {
        const lo: number = text.charCodeAt(i + 1);
        if (lo >= 0xDC00 && lo <= 0xDFFF) {
          code = 0x10000 + ((cp - 0xD800) << 10) + (lo - 0xDC00);
          i++;
        }
      }
      if (
        (code >= 0x4E00 && code <= 0x9FFF) ||   // CJK 统一表意文字
        (code >= 0x3400 && code <= 0x4DBF) ||   // 扩展 A
        (code >= 0x20000 && code <= 0x2A6DF) || // 扩展 B
        (code >= 0x2A700 && code <= 0x2B73F) || // 扩展 C/D
        (code >= 0xF900 && code <= 0xFAFF)      // 兼容汉字
      ) {
        cjk++;
      } else if (
        cp === 0xFFFD ||                        // U+FFFD 替换字符（解码失败标记）
        (cp >= 0x80 && cp <= 0x9F) ||           // C1 控制字符（乱码常见）
        (cp >= 0x202E && cp <= 0x202F)          // 右向标记等隐式控制字符
      ) {
        abnormal++;
      }
    }
    const cjkRatio: number = cjk / n;
    const abnormalRatio: number = abnormal / n;
    // CJK 占比越高越好；异常字符占比直接惩罚（最多扣到 0）
    return Math.max(0, cjkRatio - abnormalRatio * 2.0);
  }

  /** 安全解码：失败返回空串（不抛异常）。 */
  private static decodeBytes(bytes: Uint8Array, enc: string): string {
    try {
      const decoder: util.TextDecoder = new util.TextDecoder(enc);
      return decoder.decodeToString(bytes);
    } catch {
      return '';
    }
  }

  /** 字节校验是否像 UTF-8（对齐 iOS/安卓：采样尾部不完整序列视为合法前缀）。 */
  private static looksLikeUtf8(bytes: Uint8Array): boolean {
    const len: number = bytes.length;
    let i: number = 0;
    while (i < len) {
      const b: number = bytes[i];
      if (b < 0x80) {
        i++;
      } else if ((b & 0xe0) === 0xc0) {
        // 2 字节序列：续字节越界说明是采样截断的不完整前缀，视为合法（对齐 iOS/安卓）
        if (i + 1 >= len) {
          return true;
        }
        if ((bytes[i + 1] & 0xc0) !== 0x80) {
          return false;
        }
        i += 2;
      } else if ((b & 0xf0) === 0xe0) {
        // 3 字节序列
        if (i + 2 >= len) {
          return true;
        }
        if ((bytes[i + 1] & 0xc0) !== 0x80 || (bytes[i + 2] & 0xc0) !== 0x80) {
          return false;
        }
        i += 3;
      } else if ((b & 0xf8) === 0xf0) {
        // 4 字节序列
        if (i + 3 >= len) {
          return true;
        }
        if ((bytes[i + 1] & 0xc0) !== 0x80 || (bytes[i + 2] & 0xc0) !== 0x80 || (bytes[i + 3] & 0xc0) !== 0x80) {
          return false;
        }
        i += 4;
      } else {
        return false;
      }
    }
    return true;
  }
}

File: ios_app/booksclean/Core/EncodingUtil.swift
import Foundation

/// 文件编码探测工具（对齐 Android `EncodingUtil`）。
/// 探测规则：1) 检查 BOM；2) 无 BOM 时严格校验 UTF-8；3) 否则按 GB18030 兜底。
enum EncodingUtil {
    private static let DEFAULT_SAMPLE_BYTES = 8 * 1024

    /// 把编码显示名映射到 String.Encoding（iOS 原生支持 GB18030）。
    static func stringEncoding(named name: String) -> String.Encoding {
        switch name {
        case "UTF-8": return .utf8
        case "UTF-16LE": return .utf16LittleEndian
        case "UTF-16BE": return .utf16BigEndian
        case "GB18030":
            let cf = CFStringConvertEncodingToNSStringEncoding(CFStringEncoding(0x06320000)) // kCFStringEncodingGB_18030_2000
            return String.Encoding(rawValue: cf)
        default: return .utf8
        }
    }

    /// 把外部来源的 Data 复制为独立 owned 的 [UInt8]（值类型、连续内存、绝无 NSData 桥接）。
    /// 关键：security-scoped 容器文件经 FileHandle.readData 返回的 Data 可能为桥接 NSData，
    /// 对其直接做随机下标会触发 Data.subscript 的 _preconditionFailure（SIGTRAP）闪退。
    /// 此处用 withUnsafeBytes 顺序遍历复制，是唯一安全的隔离方式。
    static func toBytes(_ data: Data) -> [UInt8] {
        data.isEmpty ? [] : data.withUnsafeBytes { Array($0) }
    }

    /// 探测文件编码并返回 (编码显示名, 需跳过的 BOM 字节数)。bytes 为文件前若干字节。
    /// 入参必须为已隔离的 [UInt8]（见 toBytes），严禁直接传桥接 Data 下标访问。
    static func detectEncodingAndBom(bytes: [UInt8]) -> (String, Int) {
        let len = bytes.count
        if len >= 3 && bytes[0] == 0xEF && bytes[1] == 0xBB && bytes[2] == 0xBF { return ("UTF-8", 3) }
        if len >= 2 && bytes[0] == 0xFF && bytes[1] == 0xFE { return ("UTF-16LE", 2) }
        if len >= 2 && bytes[0] == 0xFE && bytes[1] == 0xFF { return ("UTF-16BE", 2) }
        return looksLikeUtf8(bytes) ? ("UTF-8", 0) : ("GB18030", 0)
    }

    /// 兼容封装：内部先 toBytes 隔离脏 buffer 再探测。
    static func detectEncodingAndBom(sample: Data) -> (String, Int) {
        detectEncodingAndBom(bytes: toBytes(sample))
    }

    /// 兼容封装。
    static func detectEncodingName(sample: Data) -> String { detectEncodingAndBom(sample: sample).0 }

    /// 手写 UTF-8 合法性校验；采样末尾被截断的多字节序列不算错误。
    static func looksLikeUtf8(_ b: [UInt8]) -> Bool {
        var i = 0
        let count = b.count
        while i < count {
            let c = Int(b[i])
            let need: Int
            if c < 0x80 { need = 0 }
            else if c >= 0xC2 && c <= 0xDF { need = 1 }
            else if c >= 0xE0 && c <= 0xEF { need = 2 }
            else if c >= 0xF0 && c <= 0xF4 { need = 3 }
            else { return false }
            if i + need >= count { return true }
            // 关键修复：need == 0（ASCII）时 1...0 是非法 ClosedRange，
            // Swift 运行时会触发 "Range requires lowerBound <= upperBound" fatal error（brk 1 / EXC_BREAKPOINT），
            // 表现为深度扫描读取任意含 ASCII 字节的文件即闪退。
            // 必须 guard need > 0 才进入续字节校验循环。
            if need > 0 {
                for j in 1...need {
                    if (Int(b[i + j]) & 0xC0) != 0x80 { return false }
                }
            }
            i += need + 1
        }
        return true
    }

    // MARK: - 严格解码（解决 GBK 字节被 UTF-8 lenient 解码成乱码的问题）

    /// CJK 统一表意文字（含扩展 A）占比：用于中文 txt 解码质量评分。
    /// 注意：iOS 的 `String(data:encoding:.utf8)` 对非法字节是**静默丢弃**（不插入 U+FFFD），
    /// 所以单靠 replacementCharRatio 无法识别「GBK 字节被 UTF-8 当乱码」的情况；
    /// 而正确解码的中文文本 CJK 比例应明显更高，故以 CJK 占比作为主判据。
    ///
    /// 性能：**必须遍历 `unicodeScalars` 而非 `Character`**。`String` 的 `Character` 迭代要做
    /// grapheme cluster 断字（Unicode 边界算法），对 10 万字级预览文本比标量遍历慢 1~2 个数量级，
    /// 是「预览加载时间长（主线程卡死数秒）」的主因之一。
    /// 同时按 `sampleLimit` 只抽样前若干标量评分——判别编码不需要全文，抽样即可稳定区分。
    private static func cjkScalarRatio(_ s: String, sampleLimit: Int = 20_000) -> Double {
        var cjk = 0
        var total = 0
        for v in s.unicodeScalars {
            let x = v.value
            // 跳过 ASCII 空白/换行，避免大量排版空白稀释中文占比
            if x == 0x20 || x == 0x09 || x == 0x0A || x == 0x0D { continue }
            total += 1
            if (0x3400...0x9FFF).contains(x) { cjk += 1 }
            if total >= sampleLimit { break }
        }
        return total == 0 ? 0 : Double(cjk) / Double(total)
    }

    /// 把 Data 在「字符边界安全」的前提下裁剪：丢弃末尾可能被截断的多字节序列。
    /// 预览按固定字节数截取（如 200KB）几乎必然切断一个 UTF-8/GB18030 汉字，
    /// 残字节会让 UTF-8 解码整体失败或让评分失真，表现为「预览末尾乱码 / 整篇选错编码」。
    static func trimIncompleteTail(_ bytes: [UInt8], encodingName: String) -> Data {
        let n = bytes.count
        // 保留前 keep 个字节（返回独立 Data，不再引用原 Data）
        func keep(_ count: Int) -> Data {
            count >= n ? Data(bytes) : Data(bytes[0..<max(0, count)])
        }
        if n == 0 { return Data() }
        if encodingName == "UTF-16LE" || encodingName == "UTF-16BE" {
            // UTF-16 必须按 2 字节对齐；末尾若为孤立高代理（后缺低代理），String 解码会失败，一并砍掉。
            var keepCount = (n / 2) * 2
            if keepCount >= 2 {
                let hi: Int = encodingName == "UTF-16LE"
                    ? (Int(bytes[keepCount - 2]) | (Int(bytes[keepCount - 1]) << 8))
                    : ((Int(bytes[keepCount - 2]) << 8) | Int(bytes[keepCount - 1]))
                if (0xD800...0xDBFF).contains(hi) { keepCount -= 2 } // 孤立高代理
            }
            return keepCount >= n ? Data(bytes) : Data(bytes[0..<max(0, keepCount)])
        }
        if encodingName == "UTF-8" {
            // 回退最多 3 字节找到最后一个序列起始字节，判断该序列是否完整
            var i = n - 1
            let lower = max(0, n - 4)
            while i >= lower {
                let c = bytes[i]
                if c & 0x80 == 0 { return Data(bytes) }          // ASCII 结尾，完整
                if c & 0xC0 == 0xC0 {                            // 多字节序列起始字节
                    let need: Int
                    if c & 0xE0 == 0xC0 { need = 2 }
                    else if c & 0xF0 == 0xE0 { need = 3 }
                    else if c & 0xF8 == 0xF0 { need = 4 }
                    else { return keep(i) }                      // 非法起始，直接砍掉
                    return (n - i) >= need ? Data(bytes) : keep(i)  // 字节不够即为截断
                }
                i -= 1                                           // 0x80...0xBF 续字节，继续回退
            }
            return Data(bytes)
        }
        // GB18030/GBK：双字节序列首字节 0x81...0xFE。统计末尾连续的高位字节个数，
        // 奇数说明最后一个汉字只写了一半，需要砍掉 1 字节。
        var trailing = 0
        var j = n - 1
        while j >= 0 && bytes[j] >= 0x81 && bytes[j] <= 0xFE {
            trailing += 1
            j -= 1
        }
        return trailing % 2 == 1 ? keep(n - 1) : Data(bytes)
    }

    /// 严格按候选顺序尝试解码，以「CJK 内容占比 − 异常字符惩罚」为核心判据选择最合理的解码结果。
    ///
    /// 判定逻辑（解决 iOS Foundation UTF-8 静默丢字节、不产生 U+FFFD 导致乱码的问题）：
    /// - 每个候选先做「字符边界安全裁剪」，避免尾部半个汉字影响解码与评分；
    /// - 对所有成功解码的候选计算 CJK 占比（抽样，标量级遍历）；
    /// - 真 UTF-8 中文：UTF-8 分支 CJK 高且几乎无异常字符 → 胜出；
    /// - 真 GBK/GB18030：UTF-8 分支静默丢字节后会产生大量「非 ASCII 非 CJK」的乱码字符（被惩罚），
    ///   GB18030 分支 CJK 高且异常字符少 → 选 GB18030；
    /// - 纯 ASCII/英文：各候选 CJK≈0、异常字符≈0，UTF-8 胜出（保真）。
    /// 注意：早期版本给 UTF-8 加 +0.25 红利，反而让「GBK 误当 UTF-8 静默丢字节」的乱码方案胜出，
    /// 已移除——纯按 CJK 占比与异常字符竞争即可正确区分。
    /// 返回 (解码文本, 实际使用的编码名)。全部失败返回 ("", "")。
    static func decodeStrict(data: Data, candidates: [String]) -> (String, String) {
        // 关键防护：来自 File Provider / 外部文件夹的 Data 常是桥接的 NSData（脏 buffer），
        // 其底层 length 与实际内存可能不一致，对其直接下标/复制会触发
        // Data.subscript 越界 trap（EXC_BREAKPOINT）导致闪退。
        // 必须先用 toBytes 经 withUnsafeBytes 顺序复制为独立 [UInt8]，再全部基于 [UInt8] 处理。
        let bytes = toBytes(data)
        var best = ("", "", -1.0)
        for name in candidates {
            let enc = stringEncoding(named: name)
            let safe = trimIncompleteTail(bytes, encodingName: name)
            guard let s = String(data: safe, encoding: enc), !s.isEmpty else { continue }
            var score = cjkScalarRatio(s)
            // 异常字符惩罚：非 ASCII、非 CJK 的其它字符（乱码产物）越多，越不可能是正确编码。
            let penalty = abnormalRatio(s)
            score -= penalty * 0.8
            if score > best.2 {
                best = (s, name, score)
            }
        }
        return (best.0, best.1)
    }

    /// 异常字符（乱码产物）占比：既不是 ASCII 可打印/空白、也不在常用 CJK 区的字符。
    /// GBK 文件按 UTF-8 静默丢字节解码后，会产生大量落在私用区/特殊符号区的怪字符，此处用于惩罚。
    private static func abnormalRatio(_ s: String) -> Double {
        let sampleLimit = 20000
        var total = 0
        var abnormal = 0
        for ch in s.unicodeScalars.prefix(sampleLimit) {
            total += 1
            let v = ch.value
            let isAsciiPrintable = (v >= 0x20 && v < 0x7F) || v == 0x09 || v == 0x0A || v == 0x0D
            let isCJK = (v >= 0x4E00 && v <= 0x9FFF) || (v >= 0x3400 && v <= 0x4DBF)
                       || (v >= 0xF900 && v <= 0xFAFF) || (v >= 0x3000 && v <= 0x303F)
                       || (v >= 0xFF00 && v <= 0xFFEF)
            if !isAsciiPrintable && !isCJK {
                abnormal += 1
            }
        }
        return total == 0 ? 0 : Double(abnormal) / Double(total)
    }
}
